includes/acl/sjson/sjson_parser.h

```diff
@@ -0,0 +1,719 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "sjson_parser_error.h"
+
+#include <cctype>
+#include <cmath>
+#include <cstring>
+
+namespace acl
+{
+	class SJSONParser
+	{
+	public:
+		SJSONParser(const char* const input, int input_length)
+			: m_input(input),
+			  m_input_length(input_length)
+		{
+			init_state();
+		}
+
+		bool object_begins()
+		{
+			return read_opening_brace();
+		}
+
+		bool object_begins(const char* const having_name)
+		{
+			return read_key(having_name) && read_equal_sign() && object_begins();
+		}
+
+		bool object_ends()
+		{
+			return read_closing_brace();
+		}
+
+		bool array_begins(const char* const having_name)
+		{
+			return read_key(having_name) && read_equal_sign() && read_opening_bracket();
+		}
+
+		bool try_array_begins(const char* const having_name)
+		{
+			State s = save_state();
+
+			if (!array_begins(having_name))
+			{
+				restore_state(s);
+				return false;
+			}
+
+			return true;
+		}
+
+		bool array_begins()
+		{
+			return read_opening_bracket();
+		}
+
+		bool try_array_ends() 
+		{
+			State s = save_state();
+
+			if (!array_ends())
+			{
+				restore_state(s);
+				return false;
+			}
+
+			return true;
+		}
+
+		bool array_ends()
+		{
+			return read_closing_bracket();
+		}
+
+		bool read(const char* const key, const char*& value, int& length)
+		{
+			return read_key(key) && read_equal_sign() && read_string(value, length);
+		}
+
+		bool try_read(const char* const key, const char*& value, int& length)
+		{
+			State s = save_state();
+			
+			if (!read(key, value, length))
+			{
+				restore_state(s);
+
+				value = nullptr;
+				length = 0;
+
+				return false;
+			}
+
+			return true;
+		}
+
+		bool read(const char* const key, bool& value)
+		{
+			return read_key(key) && read_equal_sign() && read_bool(value);
+		}
+
+		bool read(const char* const key, double& value)
+		{
+			return read_key(key) && read_equal_sign() && read_double(value);
+		}
+
+		bool read(const char* const key, int& value)
+		{
+			return read_key(key) && read_equal_sign() && read_int(value);
```

Comment:
all the numbers in JSON are normally possibly containing decimals, I think supporting only double and letting the caller handle invalid integer values might be fine